line-api

steward379 opened the pull request:
Line Api established and tested successfully. Ranking Vital degree by 🐶, more 🐶 more vital


-  🐶 🐶 🐶 💬 Integrate with 'Line Notify' API token in 'zuzugo' line group as test in my developer account
- Integrate with pnpm node module 'got'
-  🐶 🐶 🐶⚠️ NOTE: data type MUST be `application/x-www-form-urlencoded` instead of `multipart/form-data` at this time
- 🔍 Test spec written by Yukaii and successfully passed (& using postman to check if the API works fine)
-  🐶 👉 Wating to be done: change the arrow function in /.lib to make sure the coding style stay neat.
- **Reference:  https://github.com/aiyu666/rentHouse
-  🐶 Following above, change its language from .js to .ts (TypeScript)

steward379 committed:
line-api

Yukaii commented on pages/api/lineTest.ts:
```diff
@@ -0,0 +1,14 @@
+    return res.status(405).json({ message: "Method Not Allowed" });
+  }
+
+  sendLineNotify(`💩`, process.env.LINE_API_TOKEN);
```
可把 lineAPIToken 放到 `config.ts` 去

Yukaii replied:
message 可參考 slack 範例的值

https://github.com/Yukaii/zuzugo/blob/c491bbfc5cafe64443c5191f30e191eb49e46f23/lib/notification/slack.tsx#L22-L44

steward379 replied:
> 可把 lineAPIToken 放到 `config.ts` 去
已完成
> message 可參考 slack 範例的值
先以最基本的純文字作開展

Yukaii commented on tests/testLine.spec.ts:
```diff
@@ -0,0 +1,5 @@
+import { sendLineNotify } from "@/lib/sendLineNotify";
+
+test("Send Line Notify", async () => {
```
Line 的測試先不會在 CI 上面跑，這邊先改成：


```suggestion
test.skip("Send Line Notify", async () => {
```

steward379 replied:
> Line 的測試先不會在 CI 上面跑，這邊先改成：

done

steward379 committed:
Fix all the problems referred in the last commit, including:
- Skip the line test
- Delete the unneeded lineTest.ts
- Establish the line.ts file in @/lib/notification
    - Revise the index.ts file in @/lib/notification
- Revise some trivial syntax...

Yukaii commented on tests/testLine.spec.ts:
```diff
@@ -0,0 +1,5 @@
+import { sendLineNotify } from "@/lib/sendLineNotify";
+
+test.skip("Send Line Notify", async () => {
```
在 main branch 我有新增 houseFactory 可以產生房屋資訊的假資料，可以拿來測試

https://github.com/Yukaii/zuzugo/blob/ba45a692661d46e5ba24f0afb643aeb7c8b5d24d/tests/factory.ts#L5-L6

Yukaii replied:
記得 rebase 到 main

steward379 replied:
already rebased

Yukaii replied:
@steward379  那用假資料來送通知試試看？截張圖看看通知的結果如何~~

steward379 replied:
<img width="1125" alt="截圖 2023-02-08 17 27 53" src="https://user-images.githubusercontent.com/10606663/217489778-6f2378d9-0dff-4f96-8c3e-e3f2d392ce7c.png">

Yukaii replied:
Nice!

steward379 committed:
markdown invalid syntax revise

Yukaii commented:
因為 591 API 回傳的資料有變動，slack 訊息組合有改，Line 這邊也需要跟著修改

https://github.com/Yukaii/zuzugo/commit/f496e191183a1f3a5e1c06fdf740f0b5d506de5e

Yukaii reviewed (changes requested):
修改 Line 訊息組成

steward379 committed:
- Line test completed
    - *follow the file factory.ts to create fake data
    - *revise format a bit to comply with data types of factory.ts and faker module

steward379 committed:
- line test generated

steward379 committed:
- change line.ts similar to slack.ts Due to property changes of original data structure revision
- Redo any changes done to type.ts to avoid confusion

Yukaii commented:
@steward379 測試沒過，也先幫我改成 `test.skip` 好了

Yukaii commented on lib/notification/line.ts:
```diff
@@ -0,0 +1,45 @@
+};
+
+export async function notify(newHouses: DataItem[]) {
+  const blocks = HouseMessages({ houses: newHouses.reverse() }) as unknown as any[];
```
這邊前面幫我多加一個檢查

```ts
if (!config.tokenLine) {
  return
}
```

Yukaii commented on lib/notification/line.ts:
```diff
@@ -0,0 +1,45 @@
+/** @jsxImportSource jsx-line */
```
這邊可拿掉

```suggestion
```

Yukaii commented on lineTest.ts:
```diff
@@ -0,0 +1,17 @@
+import type { NextApiRequest, NextApiResponse } from "next";
```
lineTest 這支檔案可以移除

Yukaii commented on lib/__tests__/testLine.spec.ts:
```diff
@@ -0,0 +1,32 @@
+
+  在網頁版打開 (https://rent.591.com.tw/home/${randomHouse.post_id})
+  在 Google Maps 打開 (https://www.google.com/maps/search/?api=1&query=${randomHouse.location})
+  在手機 App 打開 (${randomHouse.mobileUrl})
```
```suggestion
  在手機 App 打開 (${mobileUrl})
```

steward379 replied:
會跳出 mobileUrl is not defined~

steward379 replied:
因為在測試裡頭並沒有直接定義該靜態網址（可在此文件增加）

moojing replied:
@Yukaii 你是這個意思嗎？https://github.com/Yukaii/zuzugo/pull/1/commits/7991ae2402ed74eb69ce2f6a7f8973fd979f3076
然後我發現 house factory 裡面的類別 `DataItem` 跟這個 factory function  裡面資料的格式好像對不太起來，
例如 `cases_name` 在 factory 有，但在 DataItem 裡面沒有這個型別。

Yukaii replied:
Sorry factory 是舊的 typing，要以 DataItem 為準

Yukaii commented on tests/factory.ts:
```diff
@@ -21,6 +21,7 @@ export function houseFactory(props: Partial<DataItem> = {}): DataItem {
     updatetime: Date.now().valueOf() / 1000,
     browsenum_all: Math.floor(Math.random() * 100000),
     location: faker.address.streetAddress(),
+    mobileURL: faker.internet.url(),
```
這行不需要

```suggestion
```

moojing replied:
removed [here](https://github.com/Yukaii/zuzugo/pull/1/commits/5f1e6f83976c5fca1654586ba5debb1ed375dd5e)

Yukaii commented on lib/notification/line.ts:
```diff
@@ -0,0 +1,45 @@
+const HouseMessages = ({ houses }: { houses: DataItem[] }) => {
+  return `
+      # 找到新租房資訊
+      ${houses.map((house) => `${house} key=${house.post_id}`)}
```
```suggestion
${houses.map((house) => `${house} key=${house.post_id}`)}
```

Yukaii commented on lib/notification/line.ts:
```diff
@@ -0,0 +1,45 @@
+  return `
+      # 找到新租房資訊
+      ${houses.map((house) => `${house} key=${house.post_id}`)}
+  `;
```
```suggestion
`;
```

Yukaii commented on lib/notification/line.ts:
```diff
@@ -0,0 +1,45 @@
+[在手機 App 打開](${mobileUrl})
+
+---
+    `;
```
`; 可以移到前一行行尾

Yukaii commented on lib/__tests__/testLine.spec.ts:
```diff
@@ -0,0 +1,32 @@
+
+  const resp = await sendLineNotify(
+    `
+  🏠 ${randomHouse.title || ""}
```
前面的空白應該都可以清理下（包含這段字串其他部分）

```suggestion
🏠 ${randomHouse.title || ""}
```

moojing replied:
整理了一下格式，不過看起來好像不支援 markdown 
https://github.com/Yukaii/zuzugo/commit/7991ae2402ed74eb69ce2f6a7f8973fd979f3076
<img width="684" alt="Screenshot 2023-02-12 at 9 31 53 PM" src="https://user-images.githubusercontent.com/11360957/218314057-c8bf9f2f-42b3-4779-b0d1-24b2192787ac.png">

steward379 replied:
的確是不支援，我這陣子測試的東西比較偏向 Line API 的其他方式，至少圖文可以分開送出（貼圖也可以，但無關緊要）。
在測試中有通過了，另外 hashtag 是可以顯示的，可以參考我們的測試 line 群。
Line API Document (messaging API)
目前我只選用 Text msg 與 Image msg
https://developers.line.biz/en/docs/messaging-api/message-types/#text-messages

steward379 replied:
[
<img width="714" alt="截圖 2023-02-13 18 04 55" src="https://user-images.githubusercontent.com/10606663/218428867-756fab8e-2709-4b04-8f7e-1e08f15ff92d.png">
](url)

Yukaii commented on lib/notification/line.ts:
```diff
@@ -0,0 +1,45 @@
+  if (!config.production) {
+    console.log(`Sending line message: ${JSON.stringify(blocks)}`);
+  }
+}
```
這邊要呼叫 `sendLineNotify` 噢，然後要確定格式是正確的

steward379 replied:
await blocks

Yukaii replied:
cc @moojing 這邊應該要改成

blocks.forEach(b => sendLineNotify(b)

之類的

Yukaii committed:
feat: init trunk

Yukaii committed:
fix: tests (#16)

Yukaii committed:
Fix branch matcher (#18)

Yukaii committed:
Setup trunk config (#17)

* feat: install trunk

* style: README

* feat: disable precommit hook

* feat: add Lint and test to required_statuses

* fix: update status check

* fix: revert the name change

* feat: more luck

* feat: more luck

* fix

* fix: setup job name explicitly

* fix: run CI for trunk-merge branches

* fix: does changing name works?

* fix: attempt

Yukaii committed:
Test: add api validation (#19)

resolve #12

Yukaii committed:
refactor: extract setup-repo action (#21)

Yukaii committed:
Feature: Check API schema reguarly (#22)

resolves #20

moojing committed:
chore: remove useless line.

moojing committed:
fix: use the right mobile url.

steward379 committed:
- Line Core Logic : Intent to use 'await' function wrapped the file 'line.ts' up.
- Line Test : adding photo messaging as the second parameter.

steward379 committed:
- complete sending line message function
- ensure sending line message including text msg & photo msg

steward379 committed:
Merge branch 'main' into line-api

steward379 committed:
fix test

Yukaii commented:
/trunk merge
